BasicRPG/BasicRPG/Program.cs

```diff
@@ -1,160 +1,121 @@
+            }
         }
-        catch (FormatException)
+        catch (FormatException) 
         {
             Console.WriteLine("Wrong input, try again!");
         }
 
-        if (characterChoice == 1)
+        switch (choice)
         {
-            Console.WriteLine($"From here on out you are {name} the mighty warrior!");
-            player = new Mage();
+            case 1: 
+                return new Warrior(name);
+            case 2:
+                return new Mage(name);   
+            default:
+                return new Idiot(name);                
         }
-        else if (characterChoice == 2)
-        {
-            Console.WriteLine($"From here on out you are {name} the cunning mage!");
-            player = new Warrior();
-        }
-
-        return player;
     }
-    static void battle(Character player, Character enemy)
+    static void Battle(Character player, Character enemy) 
     {
         Random random = new Random();
-        bool playerAlive = true;
-        bool enemyAlive = true;
-        int playerDamage;
-        int enemyDamage;
-        Console.WriteLine("You are now fighting the mighty GOBLIN!");
+        Console.WriteLine($"{player.name} is fighting {enemy.name}!");
 
-        while(playerAlive && enemyAlive)
+        while(player.IsAlive() && enemy.IsAlive()) 
         {
-            playerDamage = random.Next(1, 6);
-            enemyDamage = random.Next(1, 6);
             Console.WriteLine("Rolling the dice for player damage!");
             Console.ReadKey();
-            player.dealDamage(enemy, playerDamage);
+            player.DealDamage(enemy, random.Next(1, 6));
+            if (!enemy.IsAlive()) 
+            {
+                Console.WriteLine($"{enemy.name} has DIED! You are victorious!");
+                return;
+            }
             Console.WriteLine("Rolling the dice for enemy damage!");
             Console.ReadKey();
-            enemy.dealDamage(player, enemyDamage);
+            enemy.DealDamage(player, random.Next(1, 6));
+            if (!player.IsAlive()) 
+            {
+                Console.WriteLine($"{player.name} has DIED! You lost!");
+                return;
+            }
             Console.ReadKey();
-            playerAlive = player.isAlive();
-            enemyAlive = enemy.isAlive();
-        }
-
-        if (playerAlive == false)
-        {
-            Console.WriteLine("The player has DIED! You lost!");
-        }
-        else if (enemyAlive == false)
-        {
-            Console.WriteLine("You defeated the enemy! You are victorious!");
         }
-
-        Console.ReadKey();
-  
     }
-
 }
 
-public class Character
+public class Character 
 {
-    public bool alive;
-    public int playerHp = 20;
-    public int enemyHp = 20;
-    string name = " ";
-    public virtual void takeDamage(Character x)
-    {
+    int hp = 20;
+    public string name = "";
 
+    public Character(string name)
+    {
+        this.name = name;
     }
```

Comment:
This is a [Constructor](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/constructors). Originally, you were doing this to create instances of the character class:

```cs
Enemy enemy = new Enemy();
```

Which should just be this, since player and enemy both share the same information, so they can use the Character class.

```cs
Character enemy = new Character();
```

Now at the creation of the enemy instance, I want to initialize it's name as `"Tyler1"`. I could do this by assigning the enemy's name property after the instance has been created, like this:

```cs
Character enemy = new Character();
enemy.name = "Tyler1";
``` 

But everytime an instance is created, I expect them to have a name already. Characters in RPGs aren't supposed to go into battle with a name of "", so I've made a constructor that mandates a name to be provided and assigned to the character instance everytime you create one:

```cs
Character enemy = new Character("Tyler1");
```

Note that the constructor makes the initialization of a Character instance to have 1 parameter, which is a string called `name`.

Reply:
Moving on, you should fix the case where an empty string "" for `name` can be provided to the constructor, like this:

```cs
Character enemy = new Character("");
```

using the constructor like this will make its purpose moot. How can you ensure that an empty string is never assigned to the character's name upon instantiation?